File: app/src/androidTest/java/com/github/fo2rist/mclaren/tests/WebPreviewActivityTest.java
package com.github.fo2rist.mclaren.tests;

import android.support.test.runner.AndroidJUnit4;

import com.github.fo2rist.mclaren.pages.PreviewPage;
import com.github.fo2rist.mclaren.ui.previewscreen.PreviewActivity;
import com.github.fo2rist.mclaren.utils.ActivityTestBase;
import com.github.fo2rist.mclaren.utils.WebActivityTestRule;
import org.junit.Rule;
import org.junit.Test;
import org.junit.runner.RunWith;

import static android.support.test.espresso.assertion.ViewAssertions.doesNotExist;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.Visibility.GONE;
import static android.support.test.espresso.matcher.ViewMatchers.Visibility.VISIBLE;
import static android.support.test.espresso.matcher.ViewMatchers.withEffectiveVisibility;
import static android.support.test.espresso.web.assertion.WebViewAssertions.webContent;
import static android.support.test.espresso.web.matcher.DomMatchers.hasElementWithXpath;
import static android.support.test.espresso.web.sugar.Web.onWebView;
import static com.github.fo2rist.mclaren.testdata.FeedItems.HTML_ARTICLE_ITEM;
import static com.github.fo2rist.mclaren.utils.CustomViewAssertions.displayed;

@RunWith(AndroidJUnit4.class)
public class WebPreviewActivityTest extends ActivityTestBase {
    @Rule
    public WebActivityTestRule<PreviewActivity> rule
            = new WebActivityTestRule<>(PreviewActivity.class, false, false);
    private PreviewPage page = new PreviewPage();

    @Test
    public void testLoadsArticle() throws Exception {
        rule.launchActivity(PreviewActivity.createFeedItemIntent(context, HTML_ARTICLE_ITEM));

        onWebView()
                .check(webContent(hasElementWithXpath("/html/body/div/p/strong")))
                .check(webContent(hasElementWithXpath("/html/body/div/h2/strong")))
                .check(webContent(hasElementWithXpath("/html/body/div/table/tbody/tr/td")));
    }

    @Test
    public void testLayout() throws Exception {
        rule.launchActivity(PreviewActivity.createFeedItemIntent(context, HTML_ARTICLE_ITEM));

        page.onToolbarImage()
                .check(matches(withEffectiveVisibility(VISIBLE)));
        page.onWebView()
                .check(displayed());
        page.onGalleyView()
                .check(doesNotExist());
    }

    @Test
    public void testToolbarImageInvisibleInUrlMode() throws Exception {
        rule.launchActivity(PreviewActivity.createUrlIntent(context, "http://address_that_doesnt_exist"));

        page.onToolbarImage()
                .check(matches(withEffectiveVisibility(GONE)));
    }
}

File: app/src/androidTest/java/com/github/fo2rist/mclaren/tests/WebPreviewActivityTest.kt
package com.github.fo2rist.mclaren.tests

import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.web.assertion.WebViewAssertions.webContent
import androidx.test.espresso.web.matcher.DomMatchers.hasElementWithXpath
import androidx.test.espresso.web.webdriver.Locator
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.agoda.kakao.web.WebElementBuilder
import com.github.fo2rist.mclaren.pages.PreviewPage
import com.github.fo2rist.mclaren.testdata.FeedItems.ARTICLE_ITEM_WITH_LINKS
import com.github.fo2rist.mclaren.testdata.FeedItems.ARTICLE_ITEM_WITH_TABLES
import com.github.fo2rist.mclaren.testutilities.ActivityTestBase
import com.github.fo2rist.mclaren.testutilities.WebActivityTestRule
import com.github.fo2rist.mclaren.ui.previewscreen.PreviewActivity
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class WebPreviewActivityTest : ActivityTestBase() {
    @get:Rule
    var rule = WebActivityTestRule(PreviewActivity::class.java, false, false)
    private val page = PreviewPage()

    @Test
    fun testLayout_with_SpecificContent_in_ArticleMode() {
        rule.launchActivity(PreviewActivity.createFeedItemIntent(context, ARTICLE_ITEM_WITH_TABLES))

        page {
            webView {
                rootItem {
                    web.check(webContent(hasElementWithXpath("/html/body/div/p/strong")))
                    web.check(webContent(hasElementWithXpath("/html/body/div/h2/strong")))
                    web.check(webContent(hasElementWithXpath("/html/body/div/table/tbody/tr/td")))
                }
            }
        }
    }

    @Test
    fun testLayout_with_HeaderImages_and_WebView_in_ArticleMode() {
        rule.launchActivity(PreviewActivity.createFeedItemIntent(context, ARTICLE_ITEM_WITH_LINKS))

        page {
            toolbarImages {
                isDisplayed()
                isAtPage(0)
                swipeLeft()
                isAtPage(1)
            }
            webView {
                isDisplayed()
            }
            galleryView {
                doesNotExist()
            }
        }
    }

    @Test
    fun testLayout_without_HeaderImages_in_UrlMode() {
        rule.launchActivity(PreviewActivity.createUrlIntent(context, "http://address_that_doesnt_exist"))

        page {
            toolbarImages { isGone() }
        }
    }

    private fun WebElementBuilder.rootItem(interaction: WebElementBuilder.KWebInteraction.() -> Unit) =
            this.withElement(Locator.XPATH, "/", interaction)
}
